calcom/cal.com
# chore: recategorize apps

## What does this PR do?

This PR updates the `config.json` or `_metadata.ts` files of our app store apps, to recategorise them as per the Google Sheet link provided by in #9304.

There are also a number of supporting changes to ensure the app store & categories continue to work correctly after the recategorisation:

- Update the static list of categories (getAppCategories.ts) as per the desired set of categories
- Update the AppCategories enum, so apps can be saved in the database as per the new categories
- Replaced some static categories lists with the enum, to improve type-safety and consistency between the code & database
- Fixed the "Installed Apps" and "Admin Apps" lists improperly handling new or changed categories
  - The lists sometimes served a 404 for the new categories, due to the use of static category lists
  - The admin apps list had a bug where it didn't always show the "There are no apps in this category" messaging
- Updated the `seed-app-store.ts` file, so the set of apps that come alongside a dev environment are in-sync with the app metadata
- Replaced `seed-app-store.config.json` with loading each app's config / metadata entries directly, so its guaranteed to be in-sync with the apps we have in the codebase
- Added `syncAppMeta` cron to keep the App metadata stored in the database (categories, dirName) in sync with the raw config values.
- Added redirects for "video" category pages to "conferencing" pages

**Differences from the spreadsheet**

There were a few apps in the spreadsheet that were missing from the codebase, which I've therefore not recategorised (because there's nothing to recategorise):

- msteams (this was actually office365-video, which has been updated)
- slack
- space-booking

And a few apps that were missing from the spreadsheet, which I've recategorised as:

App | Before | After
-- | -- | --
Discord | Video | conferencing
eightxeight | video | conferencing
mirotalk | video | conferencing
sylapsvideo | video | conferencing
event-type-location-static | video | conferencing
webex | video | conferencing
zohocrm | other | crm
metapixel | analytics | (no change)
office365-video | conferencing | (no change)
rainbow | web3 | (no change)
basic | other | (no change)
booking-pages-tag | analytics | (no change)
event-type-app-card | other | (no change)
general-app-settings | other | (no change)
link-as-an-app | other | (no change)
zoho-bigin | other | (no change)

<strike>
**Potentially outstanding**

This PR does not include any data migrations to correct apps that are already in the database's `App` table.

This does mean that the old categories might still be visible and accessible from the "Apps Store" pages, as those pages infer the supported categories from the apps in the database, rather than using the static list from `getAppsCategories` or the `AppCategories` enum.

Some options we could explore here:

1. Write a data migration to update the database entries
2. Refactor the Apps Store pages to use the `getAppCategories` and load via app-store metadata, rather than what's in the database
3. Or, a hybrid of both approaches.

I don't think this is an immediate issue, because inferring the valid categories from database entries ensures forward/backward compatibility as we recategorise apps.
</strike>

Fixes #9304



**Environment**: Staging(main branch) / Production

## Type of change


- Bug fix (non-breaking change which fixes an issue)
- Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?



| Page            | See the new categories? | See apps in the right category? |
| --------------- | ----------------------- | ------------------------------- |
| Apps Store      | Yes                     | Yes                             |
| Installed Apps  | Yes                     | Yes                             |
| Categories list | Yes                     | Yes                             |
| Category page | (n/a) | Yes |
| Admin apps list | Yes                     | Yes                             |

Page | Screenshot
--- | ---
App Store | <img width="1732" alt="app store" src="https://github.com/calcom/cal.com/assets/602850/2a66522f-e7db-4d32-94d6-7dd407a43777">
Installed apps | <img width="1732" alt="installed apps 2" src="https://github.com/calcom/cal.com/assets/602850/5e00ad38-1193-4c63-953c-19acd4ba9a9b">
Categories list | <img width="1732" alt="categories list" src="https://github.com/calcom/cal.com/assets/602850/66b33caa-0214-4137-b8de-c44e2f2f74ce">
Category Page | <img width="1732" alt="category page" src="https://github.com/calcom/cal.com/assets/602850/9a6466ef-17f1-4516-a04d-b2e1712ea1ce">
Admin apps list | <img width="1732" alt="admin apps 2" src="https://github.com/calcom/cal.com/assets/602850/e6d2ce4f-016a-4f7d-a751-544e0200a5d3">

## Checklist

- [x] Ensure no breakages in tests (currently waiting for e2e tests to finish, they're taking a while!)
- [x] Add `apps/web/public/app-categorise/messaging.svg` (Job for you @PeerRich?)
- [x] If we do a data migration: Redirect `/apps/categories/video` to `/apps/categories/conferencing`?


Solves #9304
/claim #9304
